### PR TITLE
Feature/docs

### PR DESCRIPTION
**Pull request by AgwaB**
details: Add additional explanation before start section 4-1 (Infection-style dissemination mechanism)

### CODE REVIEW

File: docs/Docs.md [hunk 1]
```diff
@@ -52,9 +52,13 @@ Group member(`M_j`)의 fail을 감지하면, 노드(`M_i1`)는 Group의 다른 m
 - `Join` 메세지를 broadcast하고 각 group의 member들은 확률적으로 `reply` 한다.
 - Group별로`Join`메세지를 전용으로 처리하는 Static coordinator를 둔다.
 
+### A More Robust and Efficient SWIM
+일반적인 multicast 방식의 Dissemination은 단지 best-effort이다. 왜냐하면 패킷이 손실되면 손실로 인해 group member들 간의 상관 관계가 없는 임의의 member가 전달 될 수도 있기 때문이다. 그래서 failure detectore protocol의 membership update 정보를 `ping`, `ping-req`, `ack` 메시지에 같이 넣어서 보내는 [piggyback](https://en.wikipedia.org/wiki/Piggybacking_(data_transmission)) 방식을 사용한다. 이러한 방식의 SWIM은 constant expected 한 message overhead를 제공하며 이것은 packet 손실에 있어서 robustness하고 low latency를 제공한다.
```
Comment (zeroFruit) [inline docs/Docs.md]: > 일반적인 multicast 방식의 Dissemination은 단지 best-effort이다.

plz edit this sentence more understandable

Comment (zeroFruit) [reply docs/Docs.md]: > 이러한 방식의 SWIM은 constant expected 한 message overhead를 제공하며 이것은 packet 손실에 있어서 robustness하고 low latency를 제공한다.

Comment (AgwaB) [reply docs/Docs.md]: Modifed some parts!

**Review by zeroFruit — DISMISSED** [review]

File: docs/Docs.md [hunk 2]
```diff
@@ -36,18 +36,108 @@ SWIM failure Detector 알고리즘은 2개의 파라미터를 사용한다: `T(p
...
 새로운 노드가 들어오거나 자발적으로 나가게 되면, 해당 정보를 비슷한 방식으로 multicast한다. 그러나 어떤 노드가 Group에 들어오기 위해서는 그룹의 최소한 하나의 member를 알아야 한다. 그 방법은 다음과 같다.
 
 - Group이 well known server와 관련이 있으면, 모든 `join`은 해당 address를 통해 이루어 질 수 있다.
 - `Join` 메세지를 broadcast하고 각 group의 member들은 확률적으로 `reply` 한다.
-- Group별로`Join`메세지를 전용으로 처리하는 Static coordinator를 둔다.
\ No newline at end of file
+- Group별로`Join`메세지를 전용으로 처리하는 Static coordinator를 둔다.
+
+### A More Robust and Efficient SWIM
+일반적인 multicast 방식의 Dissemination은 패킷의 전달을 보장하지 않는 [best-effort](https://en.wikipedia.org/wiki/Best-effort_delivery) 방식이다. 왜냐하면 패킷이 손실되면 손실로 인해 group member들 간의 상관 관계가 없는 임의의 member가 전달 될 수도 있기 때문이다. 그래서 failure detectore protocol의 membership update 정보를 `ping`, `ping-req`, `ack` 메시지에 같이 넣어서 보내는 [piggyback](https://en.wikipedia.org/wiki/Piggybacking_(data_transmission)) 방식을 사용한다. SWIM은 이렇게 failure detection과 membership update dissemination component를 분리하였으며 message overhead는 constant expected 한다. 이것은 packet 손실에 있어서 robustness하고 low latency를 제공한다.
```
Comment (hihiboss) [inline docs/Docs.md]:
```suggestion
일반적인 multicast 방식의 Dissemination은 패킷의 전달을 보장하지 않는 [best-effort](https://en.wikipedia.org/wiki/Best-effort_delivery) 방식이다. 왜냐하면, 패킷의 손실로 group member 간의 상관관계가 없는 임의의 member가 전달될 수 있기 때문이다. 그래서 failure detection protocol의 membership update 정보를 `ping`, `ping-req`, `ack` 메시지에 같이 넣어서 보내는 [piggyback](https://en.wikipedia.org/wiki/Piggybacking_(data_transmission)) 방식을 사용한다. SWIM은 이렇게 failure detection과 membership update dissemination component를 분리하여 message overhead를 constant expected 하게 했다. 이것은 packet 손실에 있어서 robustness하고 low latency를 제공한다.
```

File: docs/Docs.md [hunk 3]
```diff
@@ -36,18 +36,108 @@ SWIM failure Detector 알고리즘은 2개의 파라미터를 사용한다: `T(p
...
+
+#### Implementation Hint
+
+각 그룹의 멤버 `M_i`는 버퍼를 가지고 있는데 
+
+* 버퍼에는 최근에 발생한 membership update 정보들을 담고 있다. 
+* 그리고 버퍼에 자신(`M_i`)이 주위 멤버들에게 piggyback한 횟수를 담고 있는 `local count`의 정보를 담는다. 그리고 이 `local count`로 자신이 다음에 누구에게 piggyback 할지를 결정할 수 있다. 
+* 버퍼의 사이즈가 한 멤버가 `ping` (또는 `ack`)을 piggyback 할 수 있는 최대 횟수보다 크다면(충분히 크다면) 다음으로 piggyback 할 멤버는 `local count`가 작은 멤버로 한다.  
+
+그리고 그룹 멤버 리스트를 두 개를 가지고 있다.
```
Comment (hihiboss) [inline docs/Docs.md]:
```suggestion
그리고 두 개의 그룹 멤버 리스트를 가지고 있다.
```

File: docs/Docs.md [hunk 4]
```diff
@@ -36,18 +36,108 @@ SWIM failure Detector 알고리즘은 2개의 파라미터를 사용한다: `T(p
...
+* 하나에는 아직 그룹에서 “failed” 되지 않은 멤버들을 담고 있고
+* 다른 하나에는 현재 “failed” 된 멤버들을 담고 있다.
+
+
+
+
+### Suspicion mechanism
+
+  그룹 내에 오버헤드가 큰 작업을 하고있는 프로세스 `M`이 존재할 수 있다. 그래서 이 프로세스가 `ping`을 받았지만 현재 처리하고 있는 작업때문에 제 시간에 응답을 해주지 못하였다. 그리고 그룹 내의 `M`을 제외한 다른 프로세스들은 `M`이 `ping`에 대한 응답을 받지 못했기 때문에 정상적인(non-faulty) 프로세스를 잘못됨(faulty)이라고 처리하고 네트워크에서 제외시킬 수 있다.
```
Comment (hihiboss) [inline docs/Docs.md]:
```suggestion
 만약 프로세스 `M`이 오버헤드가 큰 작업을 하고 있다면, `ping`을 받아도 오버헤드가 큰 작업 때문에 제 시간에 응답할 수 없을 것이다. 그렇다면 그룹 내의 `M`을 제외한 다른 프로세스들은 `ping`에 대한 `M`의 응답을 받지 못했기 때문에 정상 프로세스(non-faulty)를 잘못됨(faulty)으로 처리하고 네트워크에서 제외시킬 수 있다.
```

File: docs/Docs.md [hunk 5]
```diff
@@ -36,18 +36,108 @@ SWIM failure Detector 알고리즘은 2개의 파라미터를 사용한다: `T(p
...
+
+  **이렇게 non-faulty 프로세스를 faulty 프로세스로 잘못 처리하는 것을 방지**하기 위해서 SWIM은 Suspicion 메커니즘을 도입했다. Suspicion 메커니즘은 non-faulty 프로세스가 `ping`을 제시간에 응답을 보내지 못하였을 때 바로 네트워크에서 제외시키는 것이 아니라 해당 프로세스를 “suspected”라고 체크하는 것이다. 그리고 해당 프로세스가 “suspected”가 되었다고 Dissemination Component를 통해서 전파한다.
+
+이렇게 “suspected”로 선언된 프로세스가 일정시간이 지났을 때도 `ping`에 대한 응답을 보내지 못했다면 그 때 비로소 “faulty”로 해당 프로세스를 처리하고 네트워크에서 제외시킨다. 그런데 일정시간이 지나기 전에 프로세스가 `ping`에 대한 응답을 주었다면 “suspected” 된 프로세스의 상태를 “alive”로 바꾸고 해당 프로세스가 “alive” 되었다는 사실을 네트워크에 전파한다.
```
Comment (hihiboss) [inline docs/Docs.md]:
```suggestion
이렇게 “suspected”로 선언된 프로세스가 일정시간이 지나도 `ping`에 대한 응답을 보내지 못했다면 그 때 비로소 “faulty”로 해당 프로세스를 처리하고 네트워크에서 제외시킨다. 그런데 일정시간이 지나기 전에 프로세스가 `ping`에 대한 응답을 준다면 “suspected” 된 프로세스의 상태를 “alive”로 바꾸고 해당 프로세스가 “alive” 되었다는 사실을 네트워크에 전파한다.
```

File: docs/Docs.md [hunk 6]
```diff
@@ -36,18 +36,108 @@ SWIM failure Detector 알고리즘은 2개의 파라미터를 사용한다: `T(p
...
+
+#### How Suspicion Protocol Works
+
+Suspicion 프로토콜은 다음과 같이 동작한다. 예를 들어 현재 SWIM failure detector protocol period `t` 시간동안 멤버 `M_i`가 다른 멤버 `M_j`를 `ping` 타겟으로 정하고 보냈다고 생각하자. 이때 `M_i`가 `ack`를 `M_j`로부터 직접적으로나 간접적으로나 받지 못했다고 하면, `M_i`는 바로 `M_j`를 “failed”라고 선언하지 않는다. 대신 `M_j`를 “Suspected” 멤버라고 로컬 멤버십 리스트에 마크한다. 그리고
```
Comment (hihiboss) [inline docs/Docs.md]:
```suggestion
Suspicion 프로토콜은 다음과 같이 동작한다. 예를 들어 현재 SWIM failure detector protocol period `T` 시간동안 멤버 `M_i`가 다른 멤버 `M_j`를 `ping` 타겟으로 정하고 보냈다고 생각하자. 이때 `M_i`가 `M_j`에 대한 `ack`를 받지 못해도, `M_i`는 `M_j`를 바로 “failed”라고 선언하지 않는다. 대신 `M_j`를 “Suspected” 멤버라고 로컬 멤버십 리스트에 마크한다. 그리고 
```

File: docs/Docs.md [hunk 7]
```diff
@@ -36,18 +36,108 @@ SWIM failure Detector 알고리즘은 2개의 파라미터를 사용한다: `T(p
...
+
+Suspicion 프로토콜은 다음과 같이 동작한다. 예를 들어 현재 SWIM failure detector protocol period `t` 시간동안 멤버 `M_i`가 다른 멤버 `M_j`를 `ping` 타겟으로 정하고 보냈다고 생각하자. 이때 `M_i`가 `ack`를 `M_j`로부터 직접적으로나 간접적으로나 받지 못했다고 하면, `M_i`는 바로 `M_j`를 “failed”라고 선언하지 않는다. 대신 `M_j`를 “Suspected” 멤버라고 로컬 멤버십 리스트에 마크한다. 그리고 
+
+```
+{Suspect M_j: M_j suspects M_j}
+```
+
+와 같은 형태의 메시지가 Dissemination Component를 통해서 infection-style로 전파된다. 그리고 전파된 메시지를 받은 멤버 `M_i`도 `M_j`를 “Suspected”로 마크한다. 그렇지만 “Suspected” 멤버는 여전히 정상적인(non-faulty) 멤버들과 똑같이 SWIM failure detector protocol에서 `ping` 타겟으로 삼을 수 있다.
+
+protocol period `t`시간동안 `M_i`이 `M_j`에게 `ping`을 보내는데 성공했다면 로컬 멤버십 리스트에 `M_j`의 “Suspected” 마크를 지운다. 그리고
```
Comment (hihiboss) [inline docs/Docs.md]:
```suggestion
protocol period `T`시간동안 `M_i`이 `M_j`에게 `ping`을 보내는데 성공했다면 로컬 멤버십 리스트의 `M_j`에 대한 “Suspected” 마크를 지운다. 그리고 
```

File: docs/Docs.md [hunk 8]
```diff
@@ -36,18 +36,108 @@ SWIM failure Detector 알고리즘은 2개의 파라미터를 사용한다: `T(p
...
+
+```
+{Alive M_j: M_l knows M_j is alive}
+```
+
+와 같은 형태의 메시지를 Dissemination Component를 통해서 전파한다. 그리고 이런 Alive 메시지를 다른 멤버들이 받으면 똑같이 `M_j` 의 “Suspected” 마크를 지운다.
+
+`M_j` 입장에서 다른 멤버로부터 `M_j`를 “Suspected”하고 있다는 메시지를 받으면 `M_j`는 “Alive” 메시지를 다른 멤버들에게 전파할 수 있다.
+
+멤버십 리스트에서 “Suspected” 멤버 리스트는 일정 시간이 지나면 expired된다. 예를 들어 `M_h`가 `M_j`를 “Suspected” 리스트에 가지고 있었는데 일정 시간이 지날 동안 `M_j`로부터 “Alive” 메시지를 받지 못했다면 `M_h`는 `M_j`를 잘못된(faulty) 멤버라고 생각하고 로컬 멤버십 리스트에서 뺀다. 그리고 `M_j`는
```
Comment (hihiboss) [inline docs/Docs.md]:
```suggestion
멤버십 리스트에서 “Suspected” 멤버 리스트는 일정 시간이 지나면 expired된다. 예를 들어 `M_h`가 `M_j`를 “Suspected” 리스트에 가지고 있을 때, 일정 시간이 지날 동안 `M_j`로부터 “Alive” 메시지를 받지 못한다면 `M_h`는 `M_j`를 잘못된(faulty) 멤버라고 생각하고 로컬 멤버십 리스트에서 제외한다. 그리고 `M_j`는 
```

File: docs/Docs.md [hunk 9]
```diff
@@ -36,18 +36,108 @@ SWIM failure Detector 알고리즘은 2개의 파라미터를 사용한다: `T(p
...
+{Confirm M_j: M_h declares M_j as faulty}
+```
+
+와 같은 형태의 메시지를 Dissemination Component를 통해서 전파한다. 이와 같은 Suspicion 프로토콜을 통해서 failure detection의 false positives의 비율을 낮출 수 있다.
+
+#### Message Override Rule
+
+위의 예시에서 살펴보면 “Alive” 메시지는 “Suspect” 메시지를 오버라이드할 수 있다. 마찬가지로 “Confirm” 메시지는 “Suspect”메시지와 “Alive” 메시지를 오버라이드 할 수 있다. 그런데 메시지 오버라이드 규칙을 정하기위해서 메시지에 메시지 종류뿐만아니라 `incarnation number`도 추가하여야하는데 이는 멤버의 라이프사이클 동안 여러번 메시지를 받을 수 있기 때문이다. 
+
+첫 번째로 받은 “Alive” 메시지와 두 번째로 받은 “Alive” 메시지를 구분해야 하는데 이를 `incarnation number`을 통해서 하게된다. `incarnation number`는 글로벌하다. 각 멤버 `M_i`의 `incarnation number`은 그룹에 처음 들어오게되면 0으로 초기화된다. 그리고 `M_i` 스스로 Dissemination Component로부터 `M_i` 자신이 현재 incarnation에서 “Suspected” 되고 있다면 `M_i`의 `incarnation number`를 증가시키고 “Alive” 메시지를 Dissemination Component를 통해서 그룹에 전파하게된다.
```
Comment (hihiboss) [inline docs/Docs.md]:
```suggestion
첫 번째로 받은 “Alive” 메시지와 두 번째로 받은 “Alive” 메시지를 구분해야 하는데 이를 `incarnation number`을 통해서 하게된다. `incarnation number`는 글로벌하다. 각 멤버 `M_i`의 `incarnation number`은 그룹에 처음 들어오게되면 0으로 초기화된다. 그리고 Dissemination Component로부터 `M_i` 자신이 현재 incarnation에서 “Suspected” 되고 있다면 `M_i`의 `incarnation number`를 증가시키고 “Alive” 메시지를 Dissemination Component를 통해서 그룹에 전파하게된다.
```

File: docs/Docs.md [hunk 10]
```diff
@@ -36,18 +36,108 @@ SWIM failure Detector 알고리즘은 2개의 파라미터를 사용한다: `T(p
...
+* {Confirm Ml, inc=i} overrides
+    - {Alive Ml, inc=j}, any j
+    - {Alive Ml, inc=j}, any j
+```
+
+이와 같은 방식으로 각 메시지가 도착했을 때 각 멤버의 상태를 수정한다.
+
+### Round-Robin Probe Target Selection
+
+**SWIM Failure detector 알고리즘은 그룹 내에서 첫 번째로 문제가 생긴 멤버를 찾아 낼 때 큰 딜레이가 생기거나 심한 경우에는 문제가 생긴 멤버를 `ping` 타겟으로 삼지 못해서 찾아내지 못할 수도 있다.** 이러한 문제를 해결하기 위해서 각 멤버 `M_i`는 현재 멤버십에서 알고 있는 멤버들의 리스트를 들고 있고 이 중에서 `ping` 타겟을 랜덤으로 뽑는 것이 아니라 round-robin 방식으로 뽑는다. 전체 리스트를 한 바퀴 다 돌면 `M_i`는 다시 리스트를 랜덤으로 재배열한다.
```
Comment (hihiboss) [inline docs/Docs.md]:
```suggestion
**SWIM Failure detector 알고리즘은 그룹 내에서 첫 번째로 문제가 생긴 멤버를 찾아 낼 때 큰 딜레이가 생길 수 있고, 심한 경우에는 문제가 생긴 멤버를 `ping` 타겟으로 삼지 못해 멤버십의 멤버로 찾아내지 못할 수도 있다.** 이러한 문제를 해결하기 위해서 각 멤버 `M_i`는 현재 멤버십에서 알고 있는 멤버들의 리스트를 들고 있고 이 중에서 `ping` 타겟을 랜덤으로 뽑는 것이 아니라 round-robin 방식으로 뽑는다. 전체 리스트를 한 바퀴 다 돌면 `M_i`는 다시 리스트를 랜덤으로 재배열한다.
```

**Comment by zeroFruit** [comment]
plz reopen this issue. CLA recreated (#23)

File: docs/Docs.md [hunk 11]
```diff
@@ -36,18 +36,108 @@ SWIM failure Detector 알고리즘은 2개의 파라미터를 사용한다: `T(p
...
+
+이런 방식으로 `ping`을 보내게 되면 `M_i`이 가지고 있는 멤버십 리스트를 한바퀴 다 돌 때 모든 멤버들을 한 번씩은 호출하게 된다. 그래서 문제가 생긴 멤버가 생기더라도 그 멤버를 찾는데 걸리는 시간의 worst case는 `M_i`이 가지고 있는 멤버의 수이다.
```
Comment (hihiboss) [inline docs/Docs.md]:
```suggestion
이런 방식으로 `ping`을 보내게 되면 `M_i`이 가지고 있는 멤버십 리스트를 한바퀴 다 돌 때 모든 멤버들을 한 번씩은 호출하게 된다. 그래서 문제가 생긴 멤버가 생기더라도 그 멤버를 찾는데 걸리는 시간의 worst case를 `M_i`이 가지고 있는 멤버의 수로 한정할 수 있다.
```

**Comment by hihiboss** [comment]
I have commented some suggestions. I would appreciate if you refer my suggestions!

**Comment by hihiboss** [comment]
I think that we have to study it more and replace some words smoothly after merging this PR.

**Comment by zeroFruit** [comment]
@leesd556 please sign it to https://www.clahub.com/agreements/DE-labtory/swim

**Comment by zeroFruit** [comment]
plz squash commits..

`git reset --soft HEAD~<how many you commits>` may help you

this commands reset your commits but remains your works. so after reset, commit again :+1:

**Comment by AgwaB** [comment]
thx  :)

**Review by junbeomlee — APPROVED** [review]

**Pull request merged**